crypto/sphinx/src/surb.rs

```diff
@@ -0,0 +1,239 @@
+/// party knows how to decrypt the data.
+///
+/// The SURB sent to the receiving party must be accompanied
+/// by a `Pseudonym`, and once the receiving party uses that SURB to send a reply, it
+/// must be accompanied by the same `Pseudonym`.
+/// Upon receiving such a reply, the reply recipient (= sender of the SURB)
+/// uses the `Pseudonym` to find the `ReplyOpener` created with the SURB to read the reply.
+///
+/// Always use [`create_surb`] to create the [`SURB`] and [`ReplyOpener`] pair.
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct SURB<S: SphinxSuite, H: SphinxHeaderSpec> {
+    /// ID of the first relayer.
+    pub first_relayer: H::KeyId,
+    /// Alpha value for the header.
+    pub alpha: Alpha<<S::G as GroupElement<S::E>>::AlphaLen>,
+    /// Sphinx routing header.
+    pub header: RoutingInfo<H>,
+    /// Encryption key to use to encrypt the data for the SURB's creator.
+    pub sender_key: SecretKey16,
+    /// Additional data for the SURB receiver.
+    pub additional_data_receiver: H::SurbReceiverData,
+}
+
+impl<S: SphinxSuite, H: SphinxHeaderSpec> SURB<S, H> {
+    /// Size of the SURB in bytes.
+    pub const SIZE: usize = H::KEY_ID_SIZE.get()
+        + <S::G as GroupElement<S::E>>::AlphaLen::USIZE
+        + RoutingInfo::<H>::SIZE
+        + SecretKey16::LENGTH
+        + H::SURB_RECEIVER_DATA_SIZE;
+
+    /// Serializes SURB into wire format.
+    pub fn into_boxed(self) -> Box<[u8]> {
+        let alpha_len = <S::G as GroupElement<S::E>>::AlphaLen::USIZE;
+
+        let mut ret = vec![0u8; Self::SIZE];
+        ret[..H::KEY_ID_SIZE.get()].copy_from_slice(self.first_relayer.as_ref());
+        ret[H::KEY_ID_SIZE.get()..H::KEY_ID_SIZE.get() + alpha_len].copy_from_slice(self.alpha.as_ref());
+        ret[H::KEY_ID_SIZE.get() + alpha_len..H::KEY_ID_SIZE.get() + alpha_len + RoutingInfo::<H>::SIZE]
+            .copy_from_slice(self.header.as_ref());
+        ret[H::KEY_ID_SIZE.get() + alpha_len + RoutingInfo::<H>::SIZE
+            ..H::KEY_ID_SIZE.get() + alpha_len + RoutingInfo::<H>::SIZE + SecretKey16::LENGTH]
+            .copy_from_slice(self.sender_key.as_ref());
+        ret[H::KEY_ID_SIZE.get() + alpha_len + RoutingInfo::<H>::SIZE + SecretKey16::LENGTH
+            ..H::KEY_ID_SIZE.get()
+                + alpha_len
+                + RoutingInfo::<H>::SIZE
+                + SecretKey16::LENGTH
+                + H::SURB_RECEIVER_DATA_SIZE]
+            .copy_from_slice(self.additional_data_receiver.as_ref());
+
+        ret.into_boxed_slice()
+    }
+
+    /// Computes Keccak256 hash of the SURB.
+    ///
+    /// The given `context` is appended to the input.
+    pub fn get_hash(&self, context: &[u8]) -> Hash {
+        Hash::create(&[
+            self.first_relayer.as_ref(),
+            self.alpha.as_ref(),
+            self.sender_key.as_ref(),
+            self.header.as_ref(),
+            context,
+        ])
+    }
+}
```

Comment:
To make this future proof for PQ, we should reference the returned types by the `impl Hash`.

Reply:
Hashes are PQ safe mostly, we will just swap for Keccak384 or Keccak512.